airbyte-integrations/connectors/destination-bigquery/src/main/kotlin/io/airbyte/integrations/destination/bigquery/BigQueryDestination.kt

```diff
@@ -0,0 +1,394 @@
+import io.airbyte.integrations.destination.bigquery.typing_deduping.BigQuerySqlGenerator
+import io.airbyte.protocol.models.v0.AirbyteConnectionStatus
+import io.airbyte.protocol.models.v0.AirbyteMessage
+import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog
+import io.github.oshai.kotlinlogging.KotlinLogging
+import java.io.ByteArrayInputStream
+import java.io.IOException
+import java.util.*
+import java.util.function.Consumer
+import org.apache.commons.lang3.StringUtils
+
+private val log = KotlinLogging.logger {}
+
+class BigQueryDestination : BaseConnector(), Destination {
+
+    override fun check(config: JsonNode): AirbyteConnectionStatus? {
+        try {
+            val datasetId = getDatasetId(config)
+            val datasetLocation = getDatasetLocation(config)
+            val bigquery = getBigQuery(config)
+            val uploadingMethod = getLoadingMethod(config)
+
+            checkHasCreateAndDeleteDatasetRole(bigquery, datasetId, datasetLocation)
+
+            val dataset = getOrCreateDataset(bigquery, datasetId, datasetLocation)
+            if (dataset.location != datasetLocation) {
+                throw ConfigErrorException(
+                    "Actual dataset location doesn't match to location from config",
+                )
+            }
+            val queryConfig =
+                QueryJobConfiguration.newBuilder(
+                        String.format(
+                            "SELECT * FROM `%s.INFORMATION_SCHEMA.TABLES` LIMIT 1;",
+                            datasetId,
+                        ),
+                    )
+                    .setUseLegacySql(false)
+                    .build()
+
+            if (UploadingMethod.GCS == uploadingMethod) {
+                val status = checkGcsPermission(config)
+                if (status!!.status != AirbyteConnectionStatus.Status.SUCCEEDED) {
+                    return status
+                }
+            }
+
+            val result = executeQuery(bigquery, queryConfig)
+            if (result.getLeft() != null) {
+                return AirbyteConnectionStatus()
+                    .withStatus(AirbyteConnectionStatus.Status.SUCCEEDED)
+            } else {
+                throw ConfigErrorException(result.getRight())
+            }
+        } catch (e: Exception) {
+            log.error(e) { "Check failed." }
+            throw ConfigErrorException((if (e.message != null) e.message else e.toString())!!)
+        }
+    }
+
+    /**
+     * This method does two checks: 1) permissions related to the bucket, and 2) the ability to
+     * create and delete an actual file. The latter is important because even if the service account
+     * may have the proper permissions, the HMAC keys can only be verified by running the actual GCS
+     * check.
+     */
+    private fun checkGcsPermission(config: JsonNode): AirbyteConnectionStatus? {
+        val loadingMethod = config[bqConstants.LOADING_METHOD]
```

Comment:
oh this is ugly. `BigQueryUtils.getLoadingMethod` returns an enum, but here we get a jsonNode from the same field that's used by `getLoadingMethod`... Something to clean up and figure out I guess

Reply:
Yup. Didn't take a pass at cleaning the configs in this PR because BigQuery is one hairy config extraction in BigQueryUtils.